src/axom/primal/geometry/NURBSPatch.hpp

```diff
@@ -0,0 +1,2745 @@
+  }
+
+  /*!
+   * \brief Reset the knot vector in v
+   *
+   * \param [in] deg The target degree
+   * 
+   * \warning This method does NOT change the existing control points, 
+   *  i.e. does not perform degree elevation/reduction. 
+   *  Will replace existing knot vector with a uniform one.
+   *  
+   * \pre Requires deg_v < npts_v and deg >= 0
+   */
+  void setDegree_v(int deg)
+  {
+    SLIC_ASSERT(0 <= deg && deg < getNumControlPoints_v());
+
+    m_knotvec_v.makeUniform(getNumControlPoints_v(), deg);
+  }
+
+  /*!
+   * \brief Reset the knot vector and increase the number of control points
+   *
+   * \param [in] deg_u The target degree in u
+   * \param [in] deg_v The target degree in v
+   *
+   * \warning This method does NOT change the existing control points,
+   *  i.e. is not performing degree elevation or reduction.
+   * \pre Requires deg_u < npts_u and deg_v < npts_v
+   */
+  void setDegree(int deg_u, int deg_v)
+  {
+    setDegree_u(deg_u);
+    setDegree_v(deg_v);
+  }
+
+  /*!
+   * \brief Set the number control points in u
+   *
+   * \param [in] npts The target number of control points
+   * 
+   * \warning This method does NOT maintain the patch shape,
+   *  i.e. is not performing knot insertion/removal.
+   */
+  void setNumControlPoints(int npts_u, int npts_v)
+  {
+    SLIC_ASSERT(npts_u > getDegree_u());
+    SLIC_ASSERT(npts_v > getDegree_v());
+
+    m_controlPoints.resize(npts_u, npts_v);
+
+    if(isRational())
+    {
+      m_weights.resize(npts_u, npts_v);
+    }
+
+    m_knotvec_u.makeUniform(npts_u, getDegree_u());
+    m_knotvec_v.makeUniform(npts_v, getDegree_v());
+  }
+
+  /*!
+   * \brief Set the number control points in u
+   *
+   * \param [in] npts The target number of control points
+   * 
+   * \warning This method does NOT maintain the patch shape,
+   *  i.e. is not performing knot insertion/removal.
+   */
+  void setNumControlPoints_u(int npts)
+  {
+    SLIC_ASSERT(npts > getDegree_u());
+
+    m_controlPoints.resize(npts, getNumControlPoints_v());
+
+    if(isRational())
+    {
+      m_weights.resize(npts, getNumControlPoints_v());
+    }
+
+    m_knotvec_u.makeUniform(npts, getDegree_u());
+  }
+
+  /*!
+   * \brief Set the number control points in v
+   *
+   * \param [in] npts The target number of control points
+   * 
+   * \warning This method does NOT maintain the patch shape,
+   *  i.e. is not performing knot insertion/removal.
+   */
+  void setNumControlPoints_v(int npts)
+  {
+    SLIC_ASSERT(npts > getDegree_v());
+
+    m_controlPoints.resize(getNumControlPoints_u(), npts);
+
+    if(isRational())
+    {
+      m_weights.resize(getNumControlPoints_u(), npts);
+    }
+
+    m_knotvec_v.makeUniform(npts, getDegree_v());
+  }
+
+  /*!
+   * \brief Set the knot value in the u vector at a specific index
+   *
+   * \param [in] idx The index of the knot
+   * \param [in] knot The updated value of the knot
+   */
+  void setKnot_u(int idx, T knot) { m_knotvec_u[idx] = knot; }
```

Comment:
Should these check that `idx` is in range? Perhaps w/ a SLIC_ASSERT if we don't want it to affect performance in release configs

Reply:
Might also be a good idea to note that users can/should run a consistency check, e.g. for monotonicity after modifying the knots.

Reply:
The `KnotVector` class itself does these checks after calling the `[]` operator, since in my original inception the `KnotVector` class should be mostly responsible for making sure that it's valid. To your second point, I don't think there's any way (in debug mode, and assuming I didn't miss anything) to create an invalid knot vector in the first place, but we can be more clear in the documentation about what `isValidNURBS` is for.